lib/auth/session_access_test.go

```diff
@@ -0,0 +1,237 @@
+
+	hostRole.SetSessionRequirePolicies([]*types.SessionRequirePolicy{{
+		Filter:  "contains(participant.roles, \"participant\")",
+		Kinds:   []string{string(types.SSHSessionKind)},
+		Count:   2,
+		OnLeave: types.OnSessionLeaveTerminate,
+	}})
+
+	participantRole.SetSessionJoinPolicies([]*types.SessionJoinPolicy{{
+		Roles: []string{hostRole.GetName()},
+		Kinds: []string{string(types.SSHSessionKind)},
+		Modes: []string{string("*")},
+	}})
+
+	return startTestCase{
+		host:        hostRole,
+		sessionKind: types.SSHSessionKind,
+		participants: []SessionAccessContext{
+			{
+				Username: "participant",
+				Roles:    []types.Role{participantRole},
+			},
+			{
+				Username: "participant2",
+				Roles:    []types.Role{participantRole},
+			},
+		},
+		expected: true,
+	}
+}
+
+func failCountStartTestCase(t *testing.T) startTestCase {
+	hostRole, err := types.NewRole("host", types.RoleSpecV5{})
+	require.NoError(t, err)
+	participantRole, err := types.NewRole("participant", types.RoleSpecV5{})
+	require.NoError(t, err)
+
+	hostRole.SetSessionRequirePolicies([]*types.SessionRequirePolicy{{
+		Filter: "contains(participant.roles, \"participant\")",
+		Kinds:  []string{string(types.SSHSessionKind)},
+		Count:  3,
+	}})
+
+	participantRole.SetSessionJoinPolicies([]*types.SessionJoinPolicy{{
+		Roles: []string{hostRole.GetName()},
+		Kinds: []string{string(types.SSHSessionKind)},
+		Modes: []string{string("*")},
+	}})
+
+	return startTestCase{
+		host:        hostRole,
+		sessionKind: types.SSHSessionKind,
+		participants: []SessionAccessContext{
+			{
+				Username: "participant",
+				Roles:    []types.Role{participantRole},
+			},
+			{
+				Username: "participant2",
+				Roles:    []types.Role{participantRole},
+			},
+		},
+		expected: false,
+	}
+}
+
+func failFilterStartTestCase(t *testing.T) startTestCase {
+	hostRole, err := types.NewRole("host", types.RoleSpecV5{})
+	require.NoError(t, err)
+	participantRole, err := types.NewRole("participant", types.RoleSpecV5{})
+	require.NoError(t, err)
+
+	hostRole.SetSessionRequirePolicies([]*types.SessionRequirePolicy{{
+		Filter: "contains(participant.roles, \"host\")",
+		Kinds:  []string{string(types.SSHSessionKind)},
+		Count:  2,
+	}})
+
+	participantRole.SetSessionJoinPolicies([]*types.SessionJoinPolicy{{
+		Roles: []string{hostRole.GetName()},
+		Kinds: []string{string(types.SSHSessionKind)},
+		Modes: []string{string("*")},
+	}})
+
+	return startTestCase{
+		host:        hostRole,
+		sessionKind: types.SSHSessionKind,
+		participants: []SessionAccessContext{
+			{
+				Username: "participant",
+				Roles:    []types.Role{participantRole},
+			},
+			{
+				Username: "participant2",
+				Roles:    []types.Role{participantRole},
+			},
+		},
+		expected: false,
+	}
+}
+
+func TestSessionAccessStart(t *testing.T) {
+	testCases := []startTestCase{
+		successStartTestCase(t),
+		failCountStartTestCase(t),
+		failFilterStartTestCase(t),
+	}
+
+	for _, testCase := range testCases {
```

Comment:
You could move each test into a proper subtest with `t.Run(...)`.

This has the advantages of:
1. making it easier to see which test case(s) failed rather than just marking the entire test as a failure
2. allowing one to run a single test case without code changes